token/ts/package.json

```diff
@@ -6,19 +6,34 @@
     "repository": "https://github.com/solana-labs/solana-program-library",
     "license": "Apache-2.0",
     "engines": {
-        "node": ">= 10"
+        "node": ">= 14"
     },
     "publishConfig": {
         "access": "public"
     },
+    "files": [
+        "lib",
+        "src",
+        "README.md"
+    ],
     "type": "module",
     "sideEffects": false,
-    "main": "lib/index.js",
+    "main": "lib/index.cjs.js",
+    "module": "lib/index.esm.js",
     "esnext": "lib/index.js",
     "types": "lib/index.d.ts",
+    "browser": {
+        "./lib/index.cjs.js": "./lib/index.browser.cjs.js",
+        "./lib/index.esm.js": "./lib/index.browser.esm.js"
+    },
+    "browserslist": [
+        "defaults",
+        "not IE 11",
+        "maintained node versions"
+    ],
     "scripts": {
         "clean": "shx rm -rf lib",
-        "build": "yarn clean && tsc",
+        "build": "yarn clean && tsc && rollup -c",
```

Comment:
I invite you to consider not bundling the library, but just to transpile it to the typescript compilation target of your choosing (according to the presumably unused `browserslist` config above, `es5` or `es2015`, although I believe `es2017` is a great & safe target for such a library - end users in the crypto space are heavily incentivized to keep an up to date modern browser, also consumers can always re-compile, which some frameworks do by default).

Indeed, `@solana/spl-token` application consumers **are** bundlers, they will bundle it anyway, might as well allow them to retain some level of control on how it is bundled for their application.

Not bundling also brings in a number of advantages, such as:
- allow consumers to hack directly in `node_modules` and perhaps patch the library themselves locally if needed using tools such as [patch-package](https://www.npmjs.com/package/patch-package)
- improved readability
- simpler build process
- less dependencies related maintenance overhead by Solana maintainers

Seeing the `nodePolyfills()` bit in the rollup configuration, I'm guessing bundling is a way to hide these node.js only dependencies to the end consumer.
I can understand this but I believe it's not the right way, with appropriate documentation, it is trivial for the consumer to configure their bundler to include or not a polyfill for a Node.js API depending on their use case.

Providing 1-1 source/transpiled files also can help with tree-shaking, though that doesn't make a difference if the library is really side-effects free (as opposed to just the `sideEffects` hint in `package.json`.

If you may consider not bundling & just transpiling, please also consider transpiling using Babel rather than TypeScript / `tsc`, even though the cost of maintenance is higher, the control over the compilation output is much better (such as using `browserslist` with `@babel/preset-env`, which will result in smaller bundles, compiled as close as possible to the defined browser targets).

Reply:
I can't say I know what the best course here is. As much as we can try to document what they should do, lots of devs don't seem to know how to configure their bundler, and there are so many different ones with different behaviors that its a struggle to maintain.

In general I'm a fan of using ESM and not bundling/polyfilling anything, but I've seen in the wallet-adapter repo that this can lead to a lot of confusion for Next.js/Webpack users, and probably in this case non-ESM Node.js users.

> they will bundle it anyway, might as well allow them to retain some level of control on how it is bundled for their application.

Isn't this possible if their bundler uses the tsc-transpiled JS here? https://github.com/solana-labs/solana-program-library/blob/a7856b7e4e199cf47ee1625fed1b4b6434d2e2dd/token/ts/package.json#L23

I have no idea if various bundlers actually make sense of [this field](https://2ality.com/2017/06/pkg-esnext.html), or how they handle the `module`, `main`, or `browser` fields or that matter.

> please also consider transpiling using Babel rather than TypeScript / tsc

I'll try this.

Reply:
If I can add my two lamports to this. I echo similar sentiments to @0xCryptoSheik that this should remain unbundled. I'm impartial on the use of babel, but you could use TS to produce transpiled (ES6 should be a safe target) ESM and CommonJS, by having two tsconfigs. a `tsconfig.json` and `tsconfig-cjs.json`

`tsconfig-cjs.json` would be something like:

```typescript
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "CommonJS",
    "outDir": "./lib/cjs"
  },
}
```

Then you build with `tsc -p tsconfig.json && tsc -p tsconfig-cjs.json`